crates/provider-blobstore-nats/src/provider.rs

```diff
@@ -0,0 +1,391 @@
+impl NatsBlobstoreProvider {
+    pub async fn run() -> anyhow::Result<()> {
+        let host_data = load_host_data().context("failed to load host data")?;
+        let provider = Self::from_host_data(host_data);
+        let shutdown = run_provider(provider.clone(), "nats-bucket-provider")
+            .await
+            .context("failed to run provider")?;
+        let connection = get_connection();
+        let flamegraph_path = host_data
+            .config
+            .get("FLAMEGRAPH_PATH")
+            .map(String::from)
+            .or_else(|| std::env::var("PROVIDER_BLOBSTORE_NATS_FLAMEGRAPH_PATH").ok());
+        initialize_observability!("blobstore-nats-provider", flamegraph_path);
+        serve_provider_exports(
+            &connection
+                .get_wrpc_client(connection.provider_key())
+                .await?,
+            provider,
+            shutdown,
+            bindings::serve,
+        )
+        .await
+        .context("failed to serve provider exports")
+    }
+
+    /// Build a [`NatsBlobstoreProvider`] from [`HostData`]
+    pub fn from_host_data(host_data: &HostData) -> NatsBlobstoreProvider {
+        let config = NatsConnectionConfig::from_link_config(&host_data.config, &host_data.secrets);
+        if let Ok(config) = config {
+            NatsBlobstoreProvider {
+                default_config: config,
+                ..Default::default()
+            }
+        } else {
+            warn!("failed to build NATS connection configuration, falling back to default");
+            NatsBlobstoreProvider::default()
+        }
+    }
+
+    /// Attempt to connect to NATS url (with JWT credentials, if provided)
+    async fn connect(
+        &self,
+        cfg: NatsConnectionConfig,
+    ) -> anyhow::Result<async_nats::jetstream::context::Context> {
+        let mut opts = match (cfg.auth_jwt, cfg.auth_seed) {
+            (Some(jwt), Some(seed)) => {
+                let seed = KeyPair::from_seed(&seed).context("failed to parse seed key pair")?;
+                let seed = Arc::new(seed);
+                async_nats::ConnectOptions::with_jwt(jwt, move |nonce| {
+                    let seed = seed.clone();
+                    async move { seed.sign(&nonce).map_err(async_nats::AuthError::new) }
+                })
+            }
+            (None, None) => async_nats::ConnectOptions::default(),
+            _ => bail!("must provide both jwt and seed for jwt authentication"),
+        };
+        if let Some(tls_ca) = &cfg.tls_ca {
+            opts = add_tls_ca(tls_ca, opts)?;
+        } else if let Some(tls_ca_file) = &cfg.tls_ca_file {
+            let ca = fs::read_to_string(tls_ca_file)
+                .await
+                .context("failed to read TLS CA file")?;
+            opts = add_tls_ca(&ca, opts)?;
+        }
+
+        // Get the cluster_uri with proper default
+        let uri = cfg.cluster_uri.unwrap_or(DEFAULT_NATS_URI.to_string());
```

Comment:
Clippy might yell at you here
```suggestion
        // Get the cluster_uri with proper default
        let uri = cfg.cluster_uri.unwrap_or_else(|| DEFAULT_NATS_URI.to_string());
```